src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs

```diff
@@ -56,14 +60,15 @@ public enum WorkItemQueueState
     /// </summary>
     public class WorkItemQueue
     {
+        private const int spinCount = 5;
+
         private Logger log = InternalTrace.GetLogger("WorkItemQueue");
 
-        private Queue<WorkItem> _innerQueue = new Queue<WorkItem>();
-        private object _syncRoot = new object();
-#if NETCF
-        private ManualResetEvent _syncEvent = new ManualResetEvent(false);
-        private int _waitCount = 0;
-#endif
+        private readonly ConcurrentQueue<WorkItem> _innerQueue = new ConcurrentQueue<WorkItem>();
+        private readonly ManualResetEventSlim _mreAdd = new ManualResetEventSlim(false);
+
+        private int _addId = int.MinValue;
+        private int _removeId = int.MinValue;
 
```

Comment:
These ids are only a mechanism for avoiding accessing the ConcurrentQueue itself until it is appropriate to do so.  Accessing the ConcurrentQueue is relatively expensive, while a simple CAS operation is very cheap.  They have no other use then that, and could be eliminated at a performance cost.

For instance, calling TryDequeue and having it fail is significantly more expensive then simple doing the CAS on the ids.

The ids simply indicate that it is a good time to access the ConcurrentQueue.  The ConcurrentQueue could still end up doing its own CAS retries internally, or even be unable to do the operation, but it is much less likely.  Since there are no locks, another thread could still "slip through the cracks".  This does no harm, because the ConcurrentQueue itself is thread safe.